content/Improve your Rails App searches with Elasticsearch - Leandro Dominguez.md

```diff
@@ -0,0 +1,214 @@
+
+**Step 6: create your searches method in the controller, its view and routes.**
+
+My controller method looks just like this
+
+```ruby
+ def search
+   return unless query.present?
+
+   @posts = Post.our_first_query(query)
+ end
+```
+
+Also, I did this simple view with a little of Bootstrap.
+
+![Search by semantic information](images/semantic_search.png)
+
+As you can see, I created a bunch of posts using [faker]([https://github.com/faker-ruby/faker](https://github.com/faker-ruby/faker)).
+
+Take a look into that search again… Don’t you see something special? We looked for ‘_darkness_’ and although we don’t have any matching result, we can see that the ones containing ‘_dark_’ were displayed. This is because of the text analyzers. They know these two words are not the same but they do have similar semantics.
+
+Pretty easy, wasn’t it?
+
+What happens if we look for ‘_dakr_’ instead of ‘_dark_’?
+
+![No results found](images/with_typos.png)
+
+Oops, no results. Let’s fix this.
+
+**Step 7: searching with typos and highlighting the match in searches’ results.**
+
+Our search will now look like this:
+
+```ruby
+def self.highlight(query)
+ search(
+   query: {
+     multi_match: {
+       query: query,
+       fuzziness: 'AUTO',
+       fields: %w[title body]
+     }
+   },
+   highlight: {
+     pre_tags: ['<b>'],
+     post_tags: ['</b>'],
+     fields: {
+       body: {}
+     }
+   }
+ )
+end
+```
+
+In this case we are basically adding new two features: _fuzziness_ and _highlight_.
+
+With fuzziness we are making Elastic know that errors may appear in the query. This way, it will try to match similar words, so we now will not only cover semantic but syntactic errors too.
+
+With the second part, we are making the match to be highlighted by adding pre and post tags (in this case HTML bold, but you can add whatever you want).
+
+Let’s now search for ‘_dakrness_’: a word that is neither in our database nor is well written.
+
+![Final results query](images/final_result.png)
+
+## Conclusions
+
+Although in this post we’ve just covered the tip of the iceberg, we can see the potential of this search engine and how easy it is to integrate it to a Rails application. I encourage you to get lost into the wide [Elastic documentation]([https://www.elastic.co/guide/index.html](https://www.elastic.co/guide/index.html)) and discover the huge potential that it can add to our applications.
+
+We have being using Elasticsearch through some of our applications not only to admit typos and make searches smarter and faster but also to adding [synonyms to some list of words]([https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html)) in our e-commerce apps where we expect searches like ‘ps4’ to match with ‘playstation 4’, ‘playstation four’, or ‘ipad’ to match ‘i pad’ or ‘i-pad’.
```

Comment:
```suggestion
We have been using Elasticsearch in some of our applications not only to admit typos and make searches smarter and faster but also to add [synonyms to some list of words]([https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html)) in our e-commerce apps where we expect searches like ‘ps4’ to match with ‘playstation 4’, ‘playstation four’, or ‘ipad’ to match ‘i pad’ or ‘i-pad’.
```

Reply:
I feel that the conclusion started out right, but then in this last paragraph you are telling about a very specific situation, so I guess something is missing to end the article. Maybe it can be something as simple as "Feel free to reach out if you have any questions." or something else just to close the article.

Reply:
You're right. I wanted to add this and didn't know where to exactly. Let me review this. Maybe I can put it as a suggestion.